_posts/2021-02-01-Dynarec.md

```diff
@@ -0,0 +1,374 @@
+
+```
+// on compiler fetch of new page
+page = fetch_addr / PAGE_SIZE;
+if (protected[page]) return;
+mprotect(base_ram + page * PAGE_SIZE, PAGE_SIZE, PROT_READ);
+protected[page] = true;
+
+// within page fault handler (N_INSTS = PAGE_SIZE / INST_WIDTH)
+if (sig != SIGBUS && sig != SIGSEGV) return;
+page = (fault_addr - ram_base) / PAGE_SIZE;
+if (fault_addr < ram_base || page >= N_PAGES)) exit(1);
+// remove blocks from LUT (unlink blocks here if needed)
+memset(block_lut + page * N_INSTS, N_INSTS * sizeof(void*), 0);
+mprotect(ram_base + page * PAGE_SIZE, PAGE_SIZE, PROT_READ | PROT_WRITE);
+protected[page] = false;
+```
+
+The primary downside of using host memory protection for invalidation is often
+the granularity. Since writes anywhere to a page invalidate all code on that
+page, you might find that some pages are invalidated very frequently not because
+code is being frequently modified, but because there is frequently modified data
+that just happens to be located next to it. When this becomes a problem, it may
+be advantageous to switch away from host memory invalidation and instead make
+changes to the emitted code for memory accesses, perhaps doing so only when
+necessary in a manner similar to fastmem.
+
+The core of finer-grained invalidation is to keep track of precisely which
+addresses contain code, and check write addresses against it. Most often this
+takes the form of a bitmap, where the bit corresponding to an address is set by
+the compiler when it fetches the opcode there, and a whole page of bits is unset
+when the corresponding guest page is invalidated. After a write instruction is
+executed, there must be code to look up the address in this bitmap, and if that
+bit is set, invalidate the page. While accurate, doing this check for a whole
+system would take a lot of space and use a lot of overhead for most systems, but
+it can make sense when there is very limited memory, or specific pages are known
+to frequently contain both data and code. In the latter case access to those
+pages would go through a special handler, and the bitmaps would only cover that
+subset of memory.
+
+For code that truly is frequently overwritten though, even fine-grained bitmaps
+won't help. In that case the simplest option is to emit code at the beginning of
+every block that computes a checksum of the addresses it would fetch, and verify
+that this is the same as it was at compile time. If they are different, we must
+immediately exit the JIT, invalidate the block, and compile a new one. On x86,
+the hardware crc32 instructions may be useful for the checksum portion of this
+task. In many cases frequently invalidated code is also reused later, so it may
+be wise to compute the checkum up to the next branch, instead of to where the
+end of the block was at compile time. This way the checksum can be used to index
+into a backup content-based block lookup table which can replace the newly
+invalidated block without actually recompiling.
```

Comment:
When this happens, in order to reduce the code that needs to be checksummed, you can split a basic block so that the frequently overwritten code is its own BB.

Self-modifying code tends to come in two forms:
- Replacement of large number of bytes / pages at runtime
- Overwriting a single instruction very frequently

The latter benefits from the above approach.

Reply:
Do you have more details somewhere about how you distinguish the two cases you mention when detecting writes, and how you'd lookup old blocks for reuse after you split the block?

Reply:
I think I should write something specifically about invalidation in the future, since it's interesting to optimize.

Reply:
Regarding replacement of large numbers of pages, it can be useful very to hook the the system APIs which are commonly used to load new code.

For example, most "real" operating systems use mmap to map new executable pages into memory. You can hook munmap of executable pages to directly invalidate all code being unloaded (before a new mmap call might map new code at the same place).

For consoles, it might be worth having large DMAs from disk/cart check if they are overlapping executable pages before writing. Or detecting overwrite of the first protected page via a segfault and then falling back to a full overlap check.